File: 30_day_challenge_2021_January/881_boats_to_save_people.cpp
/****************************************************************
* Author            : yiorgosynkl (find me in Github: https://github.com/yiorgosynkl)
* Date created      : 20210113
* Problem link      : https://leetcode.com/problems/boats-to-save-people/
****************************************************************/

class Solution {
public:
    int numRescueBoats(vector<int> people, int limit) {
        int boats = 0, lo = 0, hi = people.size() - 1;
        sort(people.begin(), people.end()); // sort ascending
        while (lo <= hi){
            if (people[lo] + people[hi] <= limit) lo++;
            hi--; boats++;
        }
        return boats;
    }
    
    // // @lee215
    // int numRescueBoats(vector<int> people, int limit) {
    //     int i, j;
    //     sort(people.rbegin(), people.rend());
    //     for (i = 0, j = people.size() - 1; i <= j; ++i)
    //         if (people[i] + people[j] <= limit) j--;
    //     return i;
    // }
};

File: 30_day_challenge_2021_January/881_boats_to_save_people.py
################################################################
# Author            : yiorgosynkl (find me in Github: https://github.com/yiorgosynkl)
# Date created      : 20210113
# Problem link      : https://leetcode.com/problems/boats-to-save-people/
################################################################

class Solution:
    def numRescueBoats(self, people: List[int], limit: int) -> int:
        ps = sorted(people)
        boats, lo, hi = 0, 0, len(ps) - 1
        while lo <= hi:
            if lo < hi and ps[lo] + ps[hi] <= limit:
                lo += 1
            hi -= 1
            boats += 1
        return boats
                
    # @lee215
    def numRescueBoats(self, people, limit):
        people.sort(reverse=True)
        i, j = 0, len(people) - 1
        while i <= j:
            if people[i] + people[j] <= limit: j -= 1
            i += 1
        return i